### PR TITLE
Improve debugging

### PR DESCRIPTION
**Pull request by t-8ch**
This improves the out-of-the-box debugging experience, please see the commit messages for more detailed explanations.

* **Please check if the PR fulfills these requirements**
- [x] The changes have been tested locally
- [x] There are no breaking changes

### CODE REVIEW

**Comment by t-8ch** [comment]
Maybe it would also be good to distribute the .elf files in the official releases.

File: source/Makefile [hunk 1]
```diff
@@ -133,11 +133,9 @@ OPTIM=-Os -flto -finline-small-functions -findirect-inlining -fdiagnostics-color
 GLOBAL_DEFINES += $(DEV_GLOBAL_DEFS) -D USE_RTOS_SYSTICK -D MODEL_$(model) -D VECT_TAB_OFFSET=$(bootldr_size)U
 
 # Without debug code
-DEBUG=
+DEBUG=-g
```
Comment (paulfertser) [inline source/Makefile:136]: -ggdb3 please, it allows GDB to expand macros

**Review by paulfertser — COMMENTED** [review]
I think this is not enough when LTO is used. When I was last testing debugging with LTO was a pain anyway, so probably it's not that essential but fwiw I was adding this to the linker command options:
```
+            -Wl,--undefined=pxCurrentTCB \^M
+            -Wl,--undefined=pxReadyTasksLists \^M
+            -Wl,--undefined=xDelayedTaskList1 \^M
+            -Wl,--undefined=xDelayedTaskList2 \^M
+            -Wl,--undefined=pxDelayedTaskList \^M
+            -Wl,--undefined=pxOverflowDelayedTaskList \^M
+            -Wl,--undefined=xPendingReadyList \^M
+            -Wl,--undefined=xTasksWaitingTermination \^M
+            -Wl,--undefined=xSuspendedTaskList \^M
+            -Wl,--undefined=uxCurrentNumberOfTasks \^M
+            -Wl,--undefined=uxTopUsedPriority \^M
```

File: source/Makefile (hunk 1 above)
Comment (t-8ch) [reply source/Makefile]: Done. (`-g3` should do the same)

**Comment by t-8ch** [comment]
@paulfertser 
> I think this is not enough when LTO is used.

Can you elaborate what didn't work?
I didn't use that and debugging worked fine.

**Comment by t-8ch** [comment]
Distributing the ELF files with debugging info bloats the release archive from 2MiB to 14MiB.
If this is too much then it may make sense to create a dedicate debug archive.

**Comment by alvinhochun** [comment]
Hmm, here is what I am thinking: If you have the ability to debug, wouldn't you also probably be building from source to test changes? It seems a bit redundant to include the .elf files in CI builds.

**Comment by t-8ch** [comment]
It would allow to debug the actual released artifacts.
Building it manually may not use the same compiler newlib and may not exhibit the same issue that is to be debugged.

**Comment by Ralim** [comment]
To me; I would rather not have the elf in the output files as it's both large and consumes space on GitHub profile; and anyone who is really debugging things is far more likely to build locally than use the elf. 

Considering almost all bugs / issues are in our code rather than the vendor / libc code I don't see the main reason to keep all of the elf files around?


Also, given that the compiler is good at removing unused variables, I would prefer to keep it that way in the release files. Can we not use the compiler flags when performing a debug build to prevent it being removed; only when it's actually required?

I'm happy with -g3 as default as the output files are stripped of this data.

**Comment by t-8ch** [comment]
I dropped the .elf files.

We can also remove the variable. I just tried to keep the backport as close to upstream as possible.
A new version of FreeRTOS would introduce it again.

File: source/Middlewares/Third_Party/FreeRTOS/Source/tasks.c [hunk 2]
```diff
@@ -2093,6 +2100,10 @@ BaseType_t xReturn;
 	/* Prevent compiler warnings if INCLUDE_xTaskGetIdleTaskHandle is set to 0,
 	meaning xIdleTaskHandle is not used anywhere else. */
 	( void ) xIdleTaskHandle;
+
+	/* OpenOCD makes use of uxTopUsedPriority for thread debugging. Prevent uxTopUsedPriority
+	 * from getting optimized out as it is no longer used by the kernel. */
+	( void ) uxTopUsedPriority;
```
Comment (Ralim) [inline source/Middlewares/Third_Party/FreeRTOS/Source/tasks.c:2106]: Is this in upstream?

**Comment by Ralim** [comment]
I'm ok with the variable's existence, just not sure if the (void) to keep it around is upstream?
Happy to match upstream for the variable.
Otherwise update to master and happy to merge

File: source/Middlewares/Third_Party/FreeRTOS/Source/tasks.c (hunk 2 above)
Comment (t-8ch) [reply source/Middlewares/Third_Party/FreeRTOS/Source/tasks.c:2106]: Yes:
https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/534eba66ce4a5bda45d5edeeb81ac5a3cf6d0df8/tasks.c#L2108-L2110

Comment (Ralim) [reply source/Middlewares/Third_Party/FreeRTOS/Source/tasks.c:2106]: All good then.

**Comment by t-8ch** [comment]
Rebased onto current `master`.

**Comment by playday3008** [comment]
Wow, that's debuggable?

**Comment by paulfertser** [comment]
> @paulfertser
> 
> > I think this is not enough when LTO is used.
> 
> Can you elaborate what didn't work?
> I didn't use that and debugging worked fine.

With LTO what basically happens is during the link stage all the source code gets recompiled again as one huge unit, and that allows many additional optimisations in certain cases to save space. This also includes fully getting rid of unused symbols and so with LTO enabled many (all?) of those I mentioned get optimised away and so OpenOCD RTOS support can't work. Of course you won't see the issue unless you enable LTO.

File: source/Makefile [hunk 3]
```diff
@@ -132,12 +132,9 @@ OPTIM=-Os -flto -finline-small-functions -findirect-inlining -fdiagnostics-color
 # global defines ---------------------------------------------------------------
 GLOBAL_DEFINES += $(DEV_GLOBAL_DEFS) -D USE_RTOS_SYSTICK -D MODEL_$(model) -D VECT_TAB_OFFSET=$(bootldr_size)U
 
-# Without debug code
-DEBUG=
+DEBUG=-g3
```
Comment (paulfertser) [inline source/Makefile:135]: -g3 is equivalent to -ggdb3 , it's just that I personally prefer the latter as it's making obvious it's something GDB-specific.

**Comment by t-8ch** [comment]
@paulfertser I think LTO is currently always enabled in IronOS.
And OpenOCD/gdb are happy finding these symbols for me.

Did you compile with debug information?
Maybe you didn't have debug information, so by forcing these symbols into the symbol table gdb could fall back to reading the addresses from the symbol table.

File: source/Makefile (hunk 3 above)
Comment (t-8ch) [reply source/Makefile:135]: Good point.
@Ralim should I open another PR or would you just change it directly?

**Comment by paulfertser** [comment]
> @paulfertser I think LTO is currently always enabled in IronOS.

Hm, right.

> And OpenOCD/gdb are happy finding these symbols for me.

Didn't work for me when I was debugging it half a year ago, I wonder why.

> Did you compile with debug information?

Of course :)

> Maybe you didn't have debug information, so by forcing these symbols into the symbol table gdb could fall back to reading the addresses from the symbol table.

Probably that was needed because I was using -O3 (rather than -Os), not sure. Probably something else changed since then. With optimisations in LTO aggressive enough some of those symbols would be completely unavailable, having no stable address and no global name in the symbol table.

**Comment by t-8ch** [comment]
> Probably that was needed because I was using -O3 (rather than -Os)

I can't even compile with `-O3`.
```
Linking Hexfile/TS100_EN.elf
./Middlewares/Third_Party/FreeRTOS/Source/queue.c: In function 'xQueueSemaphoreTake.constprop':
./Middlewares/Third_Party/FreeRTOS/Source/tasks.c:3332:19: error: 'xTimeOut.xTimeOnEntering' may be used uninitialized in this function [-Werror=maybe-uninitialized]
 3332 |    *pxTicksToWait -= xElapsedTime;
      |                   ^
./Middlewares/Third_Party/FreeRTOS/Source/queue.c:1421:11: note: 'xTimeOut.xTimeOnEntering' was declared here
 1421 | TimeOut_t xTimeOut;
      |           ^
./Middlewares/Third_Party/FreeRTOS/Source/tasks.c:3320:5: error: 'xTimeOut.xOverflowCount' may be used uninitialized in this function [-Werror=maybe-uninitialized]
 3320 |   if( ( xNumOfOverflows != pxTimeOut->xOverflowCount ) && ( xConstTickCount >= pxTimeOut->xTimeOnEntering ) ) /*lint !e525 Indentation preferred as is to make code within pre-processor directives clearer. */
      |     ^
./Middlewares/Third_Party/FreeRTOS/Source/queue.c:1421:11: note: 'xTimeOut.xOverflowCount' was declared here
 1421 | TimeOut_t xTimeOut;
      |           ^
lto1: all warnings being treated as errors
lto-wrapper: fatal error: arm-none-eabi-g++ returned 1 exit status
compilation terminated.
/usr/lib/gcc/arm-none-eabi/10.2.0/../../../../arm-none-eabi/bin/ld: error: lto-wrapper failed
```

**Pull request merged**